### ISSUE TITLE
OOM on a simple OpenCL loop

### ISSUE THREAD

**Issue by avidov** [arrayfire/arrayfire#3112]
OOM when running this simple loop using the OpenCL backend 

```
  for (int i = 0; i < 1e6; ++i) {
    std::vector<af::array> v;
    for (int j = 0; j < 1e5; ++j) {
      v.push_back(af::randu(1000));
    }
    std::cout << "Free memory\n";
  }
```
I expect all memory to be freed, or cached at the end of the inner loop.
Same behavior with
`export AF_MAX_BUFFERS=1`

To check if we ever call native free, added a print at the native-free lmbada in DefaultMemoryManager::unlock().

src/backend/common/DefaultMemoryManager.cpp

```
void DefaultMemoryManager::unlock(void *ptr, bool user_unlock) {
    // Shortcut for empty arrays
    if (!ptr) { return; }

    // Frees the pointer outside the lock.
    uptr_t freed_ptr(nullptr, [this](void *p) {
        std::cout << "unlock(ptr=" << p  << ") nativeFree()\n";
        this->nativeFree(p);
    });
```

We never see that print, so apparently, native free is never called. Which is also evident from nvtop GPU viewer.

I'm using arrayfire v3.7.3 and Nvidia GPU
`commit 59ac7b980d1ae124aae914fb29cbf086c948954d (tag: v3.7.3)`

Distributor ID: Ubuntu
Description:    Ubuntu 18.04.5 LTS
Release:        18.04
Codename:       bionic

```
$> clinfo -l
Platform #0: NVIDIA CUDA
 +-- Device #0: Quadro GP100
 `-- Device #1: Quadro GP100
```

**Comment by avidov** [comment]
I am investigating the ArrayFire memory management in the context of a Batch-Norm layer that I wrote using ArrayFire. That layer is crashing on `Segmentation fault` (seems like stack curruption as well since it rolls many K-lines) when input is above some size but works otherwise. Note that GPU memory utilization stays  

The OOM in the issue above seems fixed on current master head  SHA d56c3bc366a593211c64318fd1151ec1dfec8059
However, it is still behaved weirdly.

The diff below add logging to the DefaultMemoryManager and add a test to array_test that shows discrpency between 2 simple cases.
The calculation in the memory manager maybe just off, but it seems correct.
It has 2 counters 

```
struct memory_info 
+        size_t allocCnt = 0;
+        size_t freeCnt = 0;
```

The counters are incremented on nativeAlloc() and nativeFree().

[possible_memory_manager_bug.txt](https://github.com/arrayfire/arrayfire/files/6218886/possible_memory_manager_bug.txt)

At the first example test we use a larger upper limit to the internal loop. The logging from the memory manager from the first test shows growing discrepancy between `allocCnt` and `freeCnt` 

```
TEST(Array, MemFreeLeak) {
  for (int i = 0; i < 1000; ++i) {
    std::vector<af::array> v;
    for (int j = 0; j < 10000; ++j) { // larger (1000) number of allocations before de-allocation of the buffer
      v.push_back(af::randu(10000));
    }
    std::cout << "Free memory\n";
  }
}
```

See the `diff` value grows 

```
DefaultMemoryManager::alloc() MiB usedUser=0.000000 (0) usedMgr=390.585938 (9999) cache=0.000000 (0) total=390.585938 (9999) allocCnt=49999 freeCnt=4 diff=49995
Free memory
...
DefaultMemoryManager::alloc() MiB usedUser=0.000000 (0) usedMgr=390.585938 (9999) cache=0.000000 (0) total=390.585938 (9999) allocCnt=59999 freeCnt=5 diff=59994
Free memory
...
DefaultMemoryManager::alloc() MiB usedUser=0.000000 (0) usedMgr=390.585938 (9999) cache=0.000000 (0) total=390.585938 (9999) allocCnt=99999 freeCnt=9 diff=99990
```

The only difference between this test and the previous one is that we use a smaller limit on the internal loop, 100 instead of 10000.

```
TEST(Array, MemFreeAll) {
  for (int i = 0; i < 1000; ++i) {
    std::vector<af::array> v;
    for (int j = 0; j < 100; ++j) {  // smaller (100) number of allocations before de-allocation of the buffer
      v.push_back(af::randu(10000));
    }
    std::cout << "Free memory\n";
  }
}
```

Note that now the `diff` value is allwats 100 

```
DefaultMemoryManager::alloc() MiB usedUser=0.000000 (0) usedMgr=3.867188 (99) cache=0.039062 (1) total=3.906250 (100) allocCnt=100 freeCnt=0 diff=100
...
Free memory
DefaultMemoryManager::alloc() MiB usedUser=0.000000 (0) usedMgr=3.867188 (99) cache=0.039062 (1) total=3.906250 (100) allocCnt=100 freeCnt=0 diff=100
...
Free memory
DefaultMemoryManager::alloc() MiB usedUser=0.000000 (0) usedMgr=3.867188 (99) cache=0.039062 (1) total=3.906250 (100) allocCnt=100 freeCnt=0 diff=100
```

**Comment by 9prady9** [comment]
@avidov thank you for reporting it. We shall take a look into it and get back to you.

Interesting that performance optimization commit changed the behaviour - change that doesn't touch allocation even indirectly.

**Comment by 9prady9** [comment]
I am assuming your original code has `unlock` although the test snippet loop with arrays doesn't.

**Comment by 9prady9** [comment]
I have tried the following code - slightly modified - used AF_MEM_INFO macro to show buffers allocated vs in-use.
```c++
        for (int i = 0; i < 100; ++i) {
            AF_MEM_INFO("Before 1e5 Loop");
            std::vector<af::array> v;
            for (int j = 0; j < 1e5; ++j) {
                v.push_back(af::randu(1000));
                AF_MEM_INFO("Each Push in Loop");
            }
            AF_MEM_INFO("After 100 Loop");
        }
```
I have tried this snippet on 3.9.0/master, 3.8.0 and 3.7.3 versions. In every test, buffers in use goes down to 0 after the inner loop is done.

I was also able to run the below code for more than couple of hours straight without going out of memory, in fact the memory always stayed within the range of 60-450 MB with all versions of arrayfire.
```c++
  for (int i = 0; i < 1e6; ++i) {
    std::vector<af::array> v;
    for (int j = 0; j < 1e5; ++j) {
      v.push_back(af::randu(1000));
    }
    std::cout << "Free memory\n";
  }
```
I had to kill the code manually as it had been running fine for more than couple of hours.

Essentially, I was not able to reproduce either out of memory issue or leak in number of buffers in use. The count of in use buffers always goes to zero after the end of the loop.

When ArrayFire's memory manager cannot accommodate the request (because all existing pooled allocations doesn't fit the requirements of new request), garbage collection is triggered. Which will try to free up resources that aren't in use. In the case where it did find some buffers to free up, `nativeFree` is called. If even then, allocation request cannot be met, then OOM is thrown. Which basically doesn't happen with the code snippet shared in the original description.
